packages/common/src/utils/personIdValidatorUtils.ts

```diff
@@ -1,29 +1,72 @@
-export const isPersonIdValid = (personId: string): boolean => {
-  const cleanPersonId = personId.replace('-', '')
-  const numbers = cleanPersonId.match(/^(\d)(\d)(\d)(\d)(\d)(\d)(\d)(\d)(\d)(\d)(\d)(\d)$/)
-  let checkSum = 0
+import { intervalToDuration, isFuture, isValid, parseISO } from 'date-fns'
 
-  if (numbers == null) {
-    return false
-  }
+const MAXIMUM_AGE = 125
 
-  let n
-  for (let i = 3; i <= 12; i++) {
-    n = parseInt(numbers[i])
-    if (i % 2 === 0) {
-      checkSum += n
-    } else {
-      checkSum += ((n * 2) % 9) + Math.floor(n / 9) * 9
-    }
-  }
+export const isPersonIdValid = (id: string): boolean => isPersonnummerValid(id) || isSamordningsnummerValid(id)
 
-  return checkSum % 10 === 0
-}
-
-export const formatPersonId = (id: string) => {
+export const formatPersonId = (id: string): string => {
   let cleanId = id.replace(/\D/g, '')
   if (cleanId.length > 8) {
     cleanId = cleanId.slice(0, 8) + '-' + cleanId.slice(8, cleanId.length)
   }
   return cleanId
 }
+
+const isPersonnummerValid = (value: string): boolean => {
+  const PERSONNUMMER_REGEXP = /^(\d{2})(\d{2})(\d{2})([0-3]\d)([-+]?)(\d{4})$/
+  const parts = PERSONNUMMER_REGEXP.exec(value)
+
+  if (!parts) return false
+
+  const parsedDate = parseISO(parts[1] + parts[2] + '-' + parts[3] + '-' + parts[4])
+
+  if (!isValid(parsedDate) || isFuture(parsedDate) || !isDateInValidRange(parsedDate)) {
+    return false
+  }
+
+  return isCheckDigitValid(parts[2] + parts[3] + parts[4] + parts[6])
+}
+
+const isSamordningsnummerValid = (value: string): boolean => {
+  const SAMORDNINGSNUMMER_REGEXP = /^(\d{2})(\d{2})(\d{2})([6-9]\d)-?(\d{4})$/
+  const parts = SAMORDNINGSNUMMER_REGEXP.exec(value)
+
+  if (!parts) return false
+
+  const day = parseInt(parts[4], 10) - 60 // 60 is the special number for samordningsnummer.
+  const parsedDate = parseISO(parts[1] + parts[2] + '-' + parts[3] + '-' + pad(day))
+
+  if (!isValid(parsedDate) || isFuture(parsedDate) || !isDateInValidRange(parsedDate)) {
+    return false
+  }
+
+  return isCheckDigitValid(parts[2] + parts[3] + parts[4] + parts[5])
+}
+
+const isCheckDigitValid = (value: string): boolean => {
```

Comment:
This function is hard to read which is understandable since it's a lot of calculations. However, if it could be split into two functions, one for calculating the check digit and one for checking, it could be easier to refactor the function in the future.

Reply:
I've actually already done that since I needed to calculate a valid check digit for the tests.